nifi-nar-bundles/nifi-jolt-record-bundle/nifi-jolt-record-processors/src/main/java/org/apache/nifi/processors/jolt/record/JoltTransformRecord.java

```diff
@@ -0,0 +1,522 @@
+                    }
+                }
+            } catch (final Exception e) {
+                getLogger().info("Processor is not valid - " + e.toString());
+                String message = "Specification not valid for the selected transformation.";
+                results.add(new ValidationResult.Builder().valid(false)
+                        .explanation(message)
+                        .build());
+            }
+        }
+
+        return results;
+    }
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onTrigger(final ProcessContext context, ProcessSession session) throws ProcessException {
+        final FlowFile original = session.get();
+        if (original == null) {
+            return;
+        }
+
+        final ComponentLog logger = getLogger();
+        final StopWatch stopWatch = new StopWatch(true);
+
+        final String transformStrategy = context.getProperty(TRANSFORM_STRATEGY).getValue();
+        final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+        final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+
+        final RecordSchema schema;
+        List<Map<String, Object>> recordList = new ArrayList<>();
+        try (final InputStream in = session.read(original);
+             final RecordReader reader = readerFactory.createRecordReader(original, in, getLogger())) {
+            schema = writerFactory.getSchema(original.getAttributes(), reader.getSchema());
+            Record record;
+
+            while ((record = reader.nextRecord()) != null) {
+                recordList.add((Map<String, Object>) DataTypeUtils.convertRecordFieldtoObject(record, RecordFieldType.RECORD.getRecordDataType(record.getSchema())));
+            }
+
+        } catch (final Exception e) {
+            logger.error("Failed to transform {}; routing to failure", new Object[]{original, e});
+            session.transfer(original, REL_FAILURE);
+            return;
+        }
+
+        final RecordSet transformedRecordSet;
+        final ClassLoader originalContextClassLoader = Thread.currentThread().getContextClassLoader();
+        try {
+            // JOLT expects arrays to be of type List where our Record code uses Object[].
+            // Make another pass of the transformed objects to change Object[] to List.
+            recordList = (List<Map<String, Object>>) normalizeJoltObjects(recordList);
+            final JoltTransform transform = getTransform(context, original);
+            if (customClassLoader != null) {
+                Thread.currentThread().setContextClassLoader(customClassLoader);
+            }
+
+            Object transformedObjects;
+            if (APPLY_TO_RECORD_SET.getValue().equals(transformStrategy)) {
+                transformedObjects = transform(transform, recordList);
```

Comment:
Could we avoid loading the data to memory in case we don't apply the transformation to the full records set? I'd expect users to, most of the time, apply a transformation on each record, don't you think?

Reply:
I waffled on whether users would be applying the transformation to each record or the whole thing, not just for processing logic but even for the default value. I will set the default to "per-record" and have a separate loop that does per-record processing in order to avoid loading the whole thing into memory.